code/mob/living/critter/flock/flockdrone.dm

```diff
@@ -370,21 +376,46 @@
 		src.antigrab_counter = 0
 	if(keys & KEY_RUN)
 		if(!src.floorrunning && isfeathertile(src.loc))
-			if(istype(src.loc, /turf/simulated/floor/feather))
+			if (length(src.grabbed_by))
+				for(var/obj/item/grab/g in src.grabbed_by)
+					if (!(g.state == GRAB_PASSIVE || g.state == GRAB_PIN)) // in the rare case you do pin a flockdrone
+						src.can_floorrun = FALSE
+						return ..()
+			else if (!src.can_floorrun)
+				src.can_floorrun = TRUE
+
+			if (istype(src.loc, /turf/simulated/floor/feather))
 				var/turf/simulated/floor/feather/floor = src.loc
+				if (floor.broken)
+					return ..()
 				if(!floor.on)
 					floor.on()
+
 			src.start_floorrunning()
 	else if(keys && src.floorrunning)
 		src.end_floorrunning()
-	. = ..()
+	return ..()
 
 /mob/living/critter/flock/drone/proc/start_floorrunning()
 	if(src.floorrunning)
 		return
 	playsound(src, "sound/misc/flockmind/flockdrone_floorrun.ogg", 50, 1, -3)
 	src.floorrunning = 1
 	src.set_density(0)
+	src.set_pulling(null)
+	if (src.pulled_by)
+		var/mob/M = src.pulled_by
+		M.set_pulling(null)
+	var/obj/item/grab/g = src.find_type_in_hand(/obj/item/grab)
+	if(g)
+		if (g.state == GRAB_PIN)
+			qdel(g)
+		else if (!istype(g, /obj/item/grab/block))
+			src.u_equip(g)
```

Comment:
why qdel pins but u_equip other grabs?

Reply:
I was mostly going off reference code for that if I remember what I was doing. I changed the second case to qdel() too if that's okay